uring/jvm/src/main/scala/fs2/io/uring/unsafe/UringSystem.scala

```diff
@@ -0,0 +1,250 @@
+        }
+      }
+  }
+
+  final class Poller private[UringSystem] (ring: UringRing) {
+
+    private[this] val sq: UringSubmissionQueue = ring.ioUringSubmissionQueue()
+    private[this] val cq: UringCompletionQueue = ring.ioUringCompletionQueue()
+
+    private[this] var pendingSubmissions: Boolean = false
+    private[this] val callbacks: Map[Short, Either[Throwable, Int] => Unit] =
+      Map.empty[Short, Either[Throwable, Int] => Unit]
+    private[this] val ids = BitSet(Short.MaxValue + 1)
+    private[this] var lastUsedId: Int = -1
+
+    private[this] def getUniqueId(): Short = {
+      val id = (lastUsedId + 1 until Short.MaxValue)
+        .find(!ids.contains(_))
+        .getOrElse(
+          (0 until lastUsedId)
+            .find(!ids.contains(_))
+            .getOrElse(
+              throw new RuntimeException("No available IDs")
+            )
+        )
+      lastUsedId = id
+      ids.add(id)
+      id.toShort
+    }
+
+    private[this] def releaseId(id: Short): Boolean = ids.remove(id.toInt)
+
+    private[UringSystem] def removeCallback(id: Short): Boolean = {
+      val removed = callbacks.remove(id).isDefined
+      if (removed) {
+        println(s"REMOVED CB WITH ID: $id")
+        println(s"CALLBACK MAP UPDATED AFTER REMOVING: $callbacks")
+        releaseId(id)
+      }
+      removed
+    }
+
+    private[UringSystem] def getSqe(cb: Either[Throwable, Int] => Unit): Short = {
+      println("GETTING SQE")
+      pendingSubmissions = true
+      val id: Short = getUniqueId()
+      callbacks.put(id, cb)
+      println(s"CALLBACK MAP UPDATED: $callbacks")
+
+      id
+    }
+
+    private[UringSystem] def getSq(): UringSubmissionQueue = sq
+
+    private[UringSystem] def close(): Unit = ring.close()
+
+    private[UringSystem] def needsPoll(): Boolean = pendingSubmissions || !callbacks.isEmpty
+
+    private[UringSystem] def poll(nanos: Long): Boolean = {
+
+      def process(
+          completionQueueCallback: UringCompletionQueueCallback
+      ): Boolean =
+        cq.process(completionQueueCallback) > 0
+
+      if (pendingSubmissions) {
+        ring.submit()
+        pendingSubmissions = false
+      }
+
+      val completionQueueCallback = new UringCompletionQueueCallback {
+        override def handle(fd: Int, res: Int, flags: Int, op: Byte, data: Short): Unit = {
+          val removedCallback = callbacks.get(data)
+
+          println(s"[HANDLE CQCB]: fd: $fd, res: $res, falgs: $flags, op: $op, data: $data")
+
+          removedCallback.foreach { cb =>
+            if (res < 0) cb(Left(new IOException("Error in completion queue entry")))
+            else cb(Right(res))
+            removeCallback(data)
+          }
+        }
+      }
+
+      if (cq.hasCompletions()) {
+        process(completionQueueCallback)
+      } else if (nanos > 0) {
+        sq.addTimeout(
+          nanos,
+          getUniqueId()
+        ) // TODO: Check why they do it in this way instead of Scala Native way
```

Comment:
I think the wait-for-CQE-with-timeout requires a newer kernel, Netty is supporting older kernels.

https://github.com/axboe/liburing/blob/f5b6c0e1d6136ddfa3e2cceff342f0f878ed1aa0/src/queue.c#L233-L237